src/ppx/graphics_util.cpp

```diff
@@ -613,40 +625,71 @@ Result CreateImageFromCompressedImage(
+    PPX_LOG_INFO("Storage size for image: " << image.size() << " bytes\n");
+    PPX_LOG_INFO("Is image compressed: " << (gli::is_compressed(image.format()) ? "YES" : "NO"));
 
-        grfx::BufferCreateInfo ci      = {};
-        ci.size                        = static_cast<uint64_t>(imageHeight) * static_cast<uint64_t>(rowStride);
-        ci.usageFlags.bits.transferSrc = true;
-        ci.memoryUsage                 = grfx::MEMORY_USAGE_CPU_TO_GPU;
+    grfx::BufferCreateInfo ci      = {};
+    ci.size                        = 0;
+    ci.usageFlags.bits.transferSrc = true;
+    ci.memoryUsage                 = grfx::MEMORY_USAGE_CPU_TO_GPU;
 
-        ppxres = pQueue->GetDevice()->CreateBuffer(&ci, &stagingBuffer);
-        if (Failed(ppxres)) {
-            return ppxres;
-        }
-        SCOPED_DESTROYER.AddObject(stagingBuffer);
+    std::vector<MipLevel> levelSizes(image.levels());
+    for (gli::texture::size_type level = 0; level < image.levels(); level++) {
+        auto& ls = levelSizes[level];
 
-        // Map and copy to staging buffer
-        void* pBufferAddress = nullptr;
-        ppxres               = stagingBuffer->MapMemory(0, &pBufferAddress);
-        if (Failed(ppxres)) {
-            return ppxres;
-        }
+        ls.width  = static_cast<uint32_t>(image.extent(level)[0]);
+        ls.height = static_cast<uint32_t>(image.extent(level)[1]);
 
-        const char* pSrc = static_cast<const char*>(image.data());
-        char*       pDst = static_cast<char*>(pBufferAddress);
-        memcpy(pDst, pSrc, image.size());
+        ls.alignedWidth  = RoundUp<uint32_t>(ls.width, grfx::GetFormatDescription(format)->bytesPerTexel);
+        ls.alignedHeight = RoundUp<uint32_t>(ls.height, grfx::GetFormatDescription(format)->bytesPerTexel);
 
-        stagingBuffer->UnmapMemory();
+        ls.rowStride        = ls.alignedWidth * grfx::GetFormatDescription(format)->bytesPerTexel;
+        ls.alignedRowStride = RoundUp<uint32_t>(ls.rowStride, rowStrideAlignment);
+
+        ls.size = static_cast<uint64_t>(ls.alignedHeight) * static_cast<uint64_t>(ls.alignedRowStride);
+
+        ls.offset = ci.size;
+        ci.size += ls.size;
     }
 
+    ppxres = pQueue->GetDevice()->CreateBuffer(&ci, &stagingBuffer);
+    if (Failed(ppxres)) {
+        return ppxres;
+    }
+    SCOPED_DESTROYER.AddObject(stagingBuffer);
+
+    // Map and copy to staging buffer
+    void* pBufferAddress = nullptr;
+    ppxres               = stagingBuffer->MapMemory(0, &pBufferAddress);
+    if (Failed(ppxres)) {
+        return ppxres;
+    }
+
+    for (gli::texture::size_type level = 0; level < image.levels(); level++) {
+        auto& ls = levelSizes[level];
+
+        const char* pSrc = static_cast<const char*>(image.data(0, 0, level));
+        char*       pDst = static_cast<char*>(pBufferAddress) + ls.offset;
+
+        if (ls.alignedRowStride == ls.rowStride) {
+            memcpy(pDst, pSrc, image.size(level));
```

Comment:
"`bytesPerTexel`" is a bit of a misleading name for the value. `grfx_format.h` actually states "For compressed formats, this field is the size required to calculate the row pitch correctly." There's not actually 4 bytes per texel if this is a compressed format - for BC7 it's actually closer to one byte per texel (or 4 bytes per 4x4 tile). For instance, given your example of a 10 x 10 image, the image is split into 9 tiles, essentially turning it into a 12 x 12 image when compressed. The metadata stores the actual dimensions of the file, and I assume the added border is ignored when decompressed.

All that goes to say that the file stride is using 12 as the width.

As a sidenote, this check itself (should) be unnecessary - I just added it in case a single memcpy would be faster than a memcpy for each line. I doubt it'll really add much overhead, though, so I'm removing the condition to make it more readable.